### DISCUSSION TITLE
When the `do_connect` Event is Triggered in SQLAlchemy for handling volatile token refresh

### DISCUSSION THREAD

**Discussion by copdips** [sqlalchemy/sqlalchemy#12133] [Usage Questions] [answered]
Hello,

I read the doc about [Generating dynamic authentication tokens for `do_connect` event](https://docs.sqlalchemy.org/en/20/core/engines.html#generating-dynamic-authentication-tokens).

After conducting some tests, I expected the `get_authentication_token()` function to be called every time `engine.connect()` is executed. This was my understanding of the `do_connect` event. However, my test results didn’t align with this expectation, which makes me think I may have misunderstood something. I hope someone can help clarify this for me. Thanks.

Below is my demo code. As you can see, the `refreshing token` message only appears twice: 
- once during the first `engine.connect()` (which is expected) 
- and again during the second `engine.connect()` after calling `engine.dispose()` (which is also expected). 

Beyond that, it doesn’t get triggered anymore.

From my understanding, the `do_connect` event should be triggered every time `engine.connect()` is called. Is my interpretation incorrect? Could you please confirm?

My ultimate goal is to ensure that, since this is an OAuth2 token, the token is verified before each query to confirm it is still valid. If the token has expired, it should be refreshed. I already have a function that handles both verification and token refreshing. Now, I need to bind this function to an event that occurs before each query execution.

here's my demo code:

```python
from snowflake.sqlalchemy import URL
from sqlalchemy import create_engine, event, text
from sqlalchemy.engine.base import Engine

engine_url_params: dict[str, str] = {
    "account": "...",
    "user": "...",
    "database": "...",
    "schema": "...",
    "warehouse": "...",
    "role": "...",
    "token": "...",
    "authenticator": "oauth",
}

engine: Engine = create_engine(URL(**engine_url_params))


def get_authentication_token():
    print(f"***Refreshing token {refresh_count}")
    return "new_token"  # the real code has a function to return the new token here


@event.listens_for(engine, "do_connect")
def provide_token_do_connect(dialect, conn_rec, cargs, cparams):
    cparams["token"] = get_authentication_token()

print("\n===case dispose engine")
connection = engine.connect()
res_params = connection.execute(text("SHOW PARAMETERS;"))
connection.close()
engine.dispose()

print("\n===case dont dispose engine")
connection = engine.connect()
res_params = connection.execute(text("SHOW PARAMETERS;"))
connection.close()

print("\n===case new connection")
for i in range(2):
    with engine.connect() as connection:
        print(f"new connection Iteration {i}")
        res = connection.execute(text("SHOW PARAMETERS;"))

print("\n===case same connection")
with engine.connect() as connection:
    for i in range(2):
        print(f"same connection Iteration {i}")
        res = connection.execute(text("SHOW PARAMETERS;"))

```

and hereunder the console ouput:

```console
===case dispose engine
***Refreshing token

===case dont dispose engine
***Refreshing token

===case new connection
same connection Iteration 0
same connection Iteration 1

===case same connection
new connection Iteration 0
new connection Iteration 1
```

**Comment by CaselIT** [comment]
Hi,

Sqlalchemy by default uses connection pooling so the dialect event do_connect happens only when a new connection is needed, not when checking out a connection from the pool.

There are other events that you can use for this, or you can disable connection pooling using nullpool.

The events are documented here https://docs.sqlalchemy.org/en/20/core/events.html

**Reply by copdips** [reply]
@CaselIT 

Thanks for your clarification. If my goal is to verify and refresh auth token on each `engine.connect()`, it seems that changing the event from `do_connect` to `checkout` do the trick.

I need to check in `conn_rec and conn_fairy` how to reset the auth token. Previously, with the `do_connect` event, the `cparams ` contained the key `token`. However, I haven’t found the same in `conn_rec` and `conn_fairy` for the `checkout` event yet."

```python
@event.listens_for(engine, "checkout")
def provide_token_do_connect(dialect, conn_rec, conn_fairy):
    get_authentication_token()
```

Now I got the below output:

```console
===case dispose engine
***Refreshing token

===case dont dispose engine
***Refreshing token

===case new connection
***Refreshing token
new connection Iteration 0
***Refreshing token
new connection Iteration 1

===case same connection
***Refreshing token
same connection Iteration 0
same connection Iteration 1
```

**Reply by CaselIT** [reply]
There are multiple ways of implementing what you want.

The easiest one that works if you know for how long the token last and that's a fixed number is to use https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.pool_recycle and only your original do_connect event.



If that's not an option then you likely need to use two events:
- the do_connect one like you are using
- a checkout event to invalidate a connection if the token has expired. You can refer to https://docs.sqlalchemy.org/en/20/core/pooling.html#more-on-invalidation and https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.checkout were it suggest you can raise a DisconnectionError to mark the connection as invalidated. that means that the connection is closed and a new one created, triggering the do_connect event

You could also use pool_pre_ping, but since you have a way of checking if the token is invalid it's likely better to use it instead of relying on an exception while checking if the connection is ok before a checkout happens. If you want to investigate it you can look at https://docs.sqlalchemy.org/en/20/core/pooling.html#pool-disconnects-pessimistic

**Reply by copdips** [reply]
@CaselIT 

From what I understand, `engine.connect()` triggers a `checkout` event, which is used to reuse an existing `connection` from the connection `pool`.

You suggested using `pool_recycle` with `do_connect`, because once a connection is recycled (often set to the duration of the token), a new connection is automatically created. At this point, the `do_connect` event is triggered.

Using `do_connect` is better than `checkout` because, in most cases, token refreshes inside the `checkout` event are unnecessary. We only need to refresh the token when it has actually expired.

Additionally, with `checkout`, we would need to validate the token during the event. If the token is expired, we would also need to invalidate the connection, and then triggers `do_connect`, which requires more code and adds unnecessary check time compared to using `pool_recycle` with `do_connect` directly. 

Is this understanding correct?

**Reply by CaselIT** [reply, marked as answer]
if you know how long the token is valid, yes the first option is simpler. Set pool_recycle to half or 2/3 or the token validity time so you know that the connection is always valid when you check it out.

If you can't know how log the token will be valid then yes, the suggestion is to use two event `do_connect` to create a connection with the token. This token would be checked for validity in checkout but there you only need to raise `DisconnectionError`. Sqlalchemy will then generate a new connection calling your `do_connect` event.
